File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def hasPathSum(self, root, targetSum):
        """
        :type root: Optional[TreeNode]
        :type targetSum: int
        :rtype: bool
        """
        if not root:
            return False
        stack=[(root,root.val)]
        while stack:
            node,val=stack.pop()
            if not node.left and not node.right:
                if val==targetSum:
                    return True
            else:
                if node.left:
                    stack.append((node.left,val+node.left.val))
                if node.right:
                    stack.append((node.right,val+node.right.val))
        return False

File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(root==nullptr) return false;

        vector<tuple<TreeNode*,int>> stack={{root,root->val}};
        
        while(!stack.empty()){
            auto [node,sum]=stack.back();
            stack.pop_back();
            if(node->left==nullptr && node->right==nullptr){
                if(sum==targetSum) return true;
            }
            if(node->left!=nullptr) stack.push_back({node->left,sum+node->left->val});

            if(node->right!=nullptr) stack.push_back({node->right,sum+node->right->val});
        }
        return false;
    }
};